File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // Agar list empty hai ya sirf ek node hai, to wahi list return karo
        if (!head || !head->next) return head;

        ListNode* prev = nullptr; // Pehle pointer null pe set karte hain
        ListNode* curr = head;    // Current pointer head pe set karte hain

        // Jab tak current null nahi ho jata, list ko reverse karte hain
        while (curr) {
            ListNode* nextNode = curr->next; // Next node ko store karte hain
            curr->next = prev;              // Current ka next pichle node pe point kare
            prev = curr;                    // Prev pointer ko current node pe le aao
            curr = nextNode;                // Current pointer ko next node pe le aao
        }

        // Last me prev pointer nayi head ban jayegi
        return prev;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // Base case: if list is empty or has one node
        if (head == null || head.next == null) {
            return head;
        }

        ListNode curr = head;
        ListNode prev = null;

        while (curr != null) {
            // Save the next node before overwriting curr.next
            ListNode temp = curr.next;
            
            // Reverse the link
            curr.next = prev;
            
            // Move pointers forward
            prev = curr;
            curr = temp;
        }

        // prev is now the new head of the reversed list
        return prev;
    }
}